dashboard/app/models/levels/dsl_defined.rb

```diff
@@ -74,7 +75,7 @@ def self.create_from_level_builder(params, level_params, old_name = nil)
       # Save updated level data to external files
       if Rails.application.config.levelbuilder_mode
         level.rewrite_dsl_file(text)
-        rewrite_i18n_file(i18n)
+        rewrite_i18n_file(i18n) if level.script_levels.any? {|sl| ScriptConstants.i18n? sl.script.name}
```

Comment:
Is the long-term plan to delete dsls.en.yml entirely and generate it during the `sync-codeorg-in` step? Or keep it for translated script levels only?

Commit 3d8eb044feb718cc1568d60ca608da84cb505c48 seems like a clear step forward, and maybe is worth merging on its own...

Reply:
I'm actually not sure!

I played around with the idea of deleting it entirely and just generating it during the sync-codeorg-in step, but ran into a couple challenges there. Specifically, that step is built entirely around the idea of being able to get all the data it needs directly from the files in the filesystem, without the need for syncing everything into a database and using ActiveRecord. Unfortunately, it turns out it's a _lot_ more complicated than I thought it would be to get all the dsls data from the filesystem. (see https://github.com/code-dot-org/code-dot-org/pull/24369#issuecomment-479282451 for more info)

So I decided to just keep it around for the translated script in https://github.com/code-dot-org/code-dot-org/pull/27885; the idea there is that this file would get updated when the levels get changed in levelbuilder (and _only_ get updated there, rather than as a part of any build process) and then consumed by the i18n sync (and _only_ get consumed there, rather than as a part of any build process).

**Unfortunately,** that approach doesn't work very well for contained levels, since it's harder for them to know whether they're in a script or not. My short-term plan at this point is just to get things working with the "keep a minimal version of dsls.en.yml around", but I don't yet know whether or not that's the long-term solution.

Reply:
I'd love to get @islemaster's input here from the Learning Platform perspective:
 - Are "contained" levels something we plan to fully-support or deprecate going forward?
 - How are these levels being found during the yearly course version clone process?

Reply:
I think @davidsbailey knows more about this than I do, but contained levels (a.k.a. prediction levels?) are used extensively in our system - I see more than 200, spread out across courses A, B, C, D, E, F, CSD, and Dance Party.  If we're planning to deprecate them for some other representation, we've got a migration ahead of us.

Reply:
> Are "contained" levels something we plan to fully-support or deprecate going forward?

yes, contained levels (a.k.a. prediction levels) are something we plan to fully support moving forward.

if some linkage could be added in the rails models so that a contained level knew what level it was contained in, would that make this problem any easier to tackle?